### PR TITLE
Several fixes related to rendering of custom components

### LINKED ISSUE #13131: [Custom Components] Blocked when using dynamic children for rendering
### Describe the bug

In gradio 6.9.0, when using dynamic children for rendering, the following error appears in the console and fails to render (the same code works in gradio 6.8.0).

Image

The svelte code:
```svelte
<script lang="ts">
  import { Gradio } from '@gradio/utils'

  const props = $props()
  // Read props to clean up the loading effect
  new Gradio(props)

  const promise = new Promise<string>((resolve) => {
    setTimeout(() => {
      resolve('hello')
    }, 500)
  })
</script>

{#await promise then value}
  <div>
    {value}
    <!-- not working -->
    {@render props.children?.()}
  </div>
{/await}

<!-- worked -->
<!-- {@render props.children?.()} -->
```

### Have you searched existing issues?  🔎

- [x] I have searched and found no existing issues

### Reproduction

Reproduction repo: https://github.com/Col0ring/gradio6.9-dynamic-children-rendering-issue

### Screenshot

_No response_

### Logs

```shell

```

### System Info

```shell
gradio version: 6.9.0
gradio_client version: 2.3.0
```

### Severity

Blocking usage of gradio

### LINKED ISSUE #13818: Custom components are remounted on every prop update (Gradio ≥ 6.20)
### Describe the bug

Found a bug where custom component are failing to load for us on the gradio-rerun component for any verison of gradio> 6.20  https://github.com/rerun-io/gradio-rerun-viewer

with an agent came up with  the following fix

```
js/core/src/MountCustomComponent.svelte


<script lang="ts">
	import { untrack } from "svelte";

	let { node, children, ...rest } = $props();

	let component = $derived(await node.component);
@@ -20,14 +22,22 @@
		const _props = node.props.props;
		const _runtime = runtime;

		const mounted = _runtime.mount(component.default, {
			target: el,
			props: {
				shared_props: _shared_props,
				props: _props,
				children
			}
		});
		// The custom component runs in its own Svelte runtime instance, so
		// its synchronous init cannot reset *this* runtime's tracking
		// context. Without untrack(), every core `$state` read the component
		// makes while mounting (e.g. the `Gradio` class copying
		// shared_props/props) becomes a dependency of this effect, and every
		// later prop update unmounts and remounts the component.
		const mounted = untrack(() =>
			_runtime.mount(component.default, {
				target: el,
				props: {
					shared_props: _shared_props,
					props: _props,
					children
				}
			})
		);

		return () => {
			_runtime.unmount(mounted);
```

validated this fixed things for me

closet issue #13131 and the fix looks like #13817 

but it doesn't look like this is a fix for this particular issue we have

### Have you searched existing issues?  🔎

- [x] I have searched and found no existing issues

### Reproduction

```bash
pip install "gradio==6.26.0" build
gradio cc create RemountProbe --template SimpleTextbox --no-configure-metadata --directory remountprobe
```
In `remountprobe/frontend/Index.svelte`, after `const gradio = new Gradio(...)`, add:
```svelte
onMount(() => { console.log("PROBE mounted"); return () => console.log("PROBE unmounted"); });
```
```bash
cd remountprobe && gradio cc build --no-generate-docs && cd .. && pip install -e remountprobe
```
```python
import time
import gradio as gr
from gradio_remountprobe import RemountProbe

def stream():
    for i in range(5):
        time.sleep(0.3)
        yield f"value-{i}"

with gr.Blocks() as demo:
    out = RemountProbe(label="probe", value="init")
    gr.Button("Go").click(stream, outputs=out)

demo.launch()
```
Click **Go**. Expected: one `PROBE mounted`. Actual: 15 `unmounted`/`mounted` pairs; the component's DOM is recreated each time.

### Screenshot

Image

### Logs

```shell

```

### System Info

```shell
gradio 6.26.0, gradio_client 2.6.1, Python 3.11, macOS, Chrome 152
Same MountCustomComponent.svelte on 6.9.0 through 6.26.0 and main
```

### Severity

Blocking usage of gradio

### PR DESCRIPTION
**Pull request by abidlabs**
## Description

Fixes custom-component regressions caused by Gradio and custom components using separate Svelte runtimes:

- Custom component children rendered after an asynchronous Svelte boundary failed because the host-runtime `children` snippet was invoked from the custom runtime without an active host effect.
- Reading Gradio's reactive prop proxies while the foreign runtime mounted registered those reads against the host `$effect`, causing the whole custom component to unmount and mount again on every prop update.
- Lazy rendering replaces the `shared_props` object when revealing an inactive tab. Tracking that object reference caused an immediate extra unmount/mount cycle every time the custom component appeared.
- `@gr.render` can reuse a custom-component mount for a node with a new ID. Without tracking that ID, the component remained registered under the old ID instead of remounting.
- Nested custom components lost Svelte context because every custom component was mounted as a separate root without receiving the ancestor custom component's context map.
- The persistent custom-component mount target became a flex/grid item, so a custom component's own root could not participate directly in its parent's layout.

This change recreates the outer child snippet with the custom component's runtime and mounts a small host-runtime wrapper inside it. It also calls the foreign runtime's mount inside the host runtime's `untrack`, reads the prop-container references without subscribing the mount effect to them, tracks the node ID so reused `@gr.render` mounts register under the new ID, exports the required runtime helpers from newly built custom component bundles, carries the ancestor custom component's context through the host wrapper into nested custom-component mounts, and uses `display: contents` on the persistent mount target so it creates no layout box.

Closes: #13131
Closes: #13818

## Reproduction

For #13131, rendering children after an async boundary failed:

```svelte
{#await promise then value}
	<div>
		{value}
		{@render children?.()}
	</div>
{/await}
```

Before this change, the delayed block throws `Cannot read properties of null (reading 'nodes')` and never renders its child. After this change, both the resolved value and child render.

For #13818, a streaming update to a custom component caused its Svelte component to unmount and mount again for every yielded value. After this change, the component remains mounted while its value updates.

For lazy-rendered custom components, opening or revisiting an inactive tab mounted, immediately unmounted, and mounted the component again. After this change, the first reveal performs one mount; revisiting the tab and streamed prop updates do not remount it.

For `@gr.render`, reusing the Svelte mount instance for a new node ID left the custom component registered under its old ID. The mount effect now tracks `node.id`, so an ID change remounts the component and registers its callbacks under the new ID.

For nested custom components, `getContext()` in a child custom component could not see values set by an ancestor custom component. The regression test mounts the provider and consumer with two independent custom-component runtimes, plus Gradio's host runtime, and verifies that the context value reaches the nested consumer.

For nested layout, custom-component roots with `flex` sizing were wrapped in a persistent `<span>` that became the parent's flex item. The mount target now uses `display: contents`, allowing the custom-component roots to participate directly in flex and grid layout.

## Before/after Spaces

One shared pair covers both #13131 and #13818 plus the nested-context, parent-layout, and duplicate lazy-mount concerns raised in review. The two Spaces run the same comprehensive custom-component showcase; the before Space uses the released Gradio package, while the after Space uses the preview wheel built from this PR's current head.

- [Before fix](https://huggingface.co/spaces/abidlabs/gradio-13131-before) — latest released Gradio; all regressions are visible.
- [After fix](https://huggingface.co/spaces/abidlabs/gradio-13131-after) — preview wheel from this PR; all scenarios work.

Use the four tabs to verify:

- **Children and layout:** synchronous children still render, asynchronous children resolve, and the expected direct-child layout is retained.
- **Nested flex layout:** two nested custom-component cards with deliberately different content should be equal-width columns with a 16 px gap. Before the fix they are content-sized because their mount wrappers become flex items; after the fix both card roots participate directly in flex sizing.
- **Nested context:** a standalone consumer receives no context, a direct nested consumer receives `outer-context`, and a deeply nested consumer receives `inner-context`.
- **Reactive updates:** streaming reaches `value-4`; before the fix the mount count increases from 2 to 17. After the fix the first reveal starts at 1, and neither streaming nor switching away and back adds mounts.

Both Spaces use an identical `app.py` (`SHA-256: 56bde398420f8f4511fe99d19584efa71eff8f8b0920dabb69a2159c5db32fe7`).

## Testing

- `CI=1 pnpm exec vitest run js/core/src/MountCustomComponent.test.ts --config .config/vitest.config.ts --reporter=verbose` (5 tests passed)
- focused Prettier check for all changed frontend files
- focused ESLint check for all changed frontend source files
- `pnpm --filter @gradio/preview build`
- `pnpm --filter @self/spa build`
- manually verified the detailed #13131 reproduction linked in review: after updating its stale dependencies and adding the stylesheet its current loader expects, the delayed `hello` value and nested `Run` button both render
- manually verified #13818 with a streaming custom component: the unpatched frontend remounted repeatedly; the patched frontend reached `value-4` without any additional mount/unmount logs
- manually verified lazy tab rendering: the patched frontend performs one mount on first reveal and does not remount when revisiting the tab
- manually verified the nested flex-layout Space at a 1280 px viewport: before, the cards measured 119 px and 692 px; after, both measured 583 px with the intended 16 px gap

`bash scripts/format_frontend.sh` formatted the changes successfully. Its repository-wide `svelte-check` phase currently reports unrelated existing errors in the dispatcher callback type, PrismJS imports, preview Rollup types, and ColorPicker.

## AI Disclosure

- [x] I used AI to reproduce and diagnose the issues, draft the implementation and tests, and prepare this PR description. I reviewed every changed line and ran the checks listed above.
- [ ] I did not use AI

## 🎯 PRs Should Target Issues

This PR targets and closes #13131 and #13818. I checked both issues for overlapping open PRs before adding their fixes and found none.

## Testing and Formatting Your Code

The frontend formatting and verification details are listed above. No backend code was changed.

### CODE REVIEW

**Comment by abidlabs** [comment]
## Shared custom-component before/after verification

One Space pair covers #13131, #13818, and the nested-context, parent-layout, and duplicate lazy-mount concerns raised in review. Both sides use the same expanded `app.py` (SHA-256 `56bde398420f8f4511fe99d19584efa71eff8f8b0920dabb69a2159c5db32fe7`) and custom-component assets; the only Gradio difference is the latest release versus this PR's preview wheel.

- **Before:** https://huggingface.co/spaces/abidlabs/gradio-13131-before
- **After:** https://huggingface.co/spaces/abidlabs/gradio-13131-after

| Tab | Before — latest release | After — PR wheel `0179ea538` |
| --- | --- | --- |
| Children and layout | Synchronous children render, but the async card remains at `Waiting for the custom component…` and its children never appear. | Synchronous and delayed children render. The async grid contains the button and textbox as two direct DOM children. |
| Nested flex layout | The short and long nested custom-component cards are content-sized (119 px and 692 px at a 1280 px viewport) because their mount `<span>` elements become the flex items. | Both card roots participate directly in flex sizing and measure 583 px, with the intended 16 px gap; the persistent mount targets compute to `display: contents`. |
| Nested context | Standalone, direct-nested, and deep-nested consumers all show `No context received`. | The standalone control has no context; nested consumers show `outer-context` and the nearest-provider override `inner-context`. |
| Reactive updates | Streaming reaches `value-4`, but the visible mount count increases from 2 to 17. Opening or revisiting the tab also adds mounts in pairs. | The first reveal starts at mount count 1. Streaming reaches `value-4`, and neither streaming nor switching away and back increases the count. |

Open the pair side by side and work through all four tabs. The **Nested flex layout** tab specifically demonstrates Col0ring's review concern: the mount target remains persistent for reactive updates without creating a flex/grid layout box. The **Reactive updates** tab now also demonstrates that lazy rendering does not trigger an immediate duplicate mount.

The focused unit suite additionally covers `@gr.render` reusing the Svelte mount instance for a new node ID: the custom component remounts once and registers its callbacks under that new ID.

File: js/core/src/MountCustomComponent.svelte [hunk 1]
```diff
@@ -19,13 +23,31 @@
 		const _shared_props = node.props.shared_props;
 		const _props = node.props.props;
 		const _runtime = runtime;
+		// Recreate the snippet in the custom component's Svelte runtime, then
+		// mount the original children with Gradio's runtime inside its placeholder.
+		const runtime_children =
+			children && _runtime.createRawSnippet
+				? _runtime.createRawSnippet(() => ({
+						render: () => "<span></span>",
```
[the remark this answers is not in the document]
Comment (abidlabs) [reply js/core/src/MountCustomComponent.svelte]: Addressed in ed5009942. The foreign-runtime raw snippet now creates only a temporary hidden marker; after the host runtime mounts into it, target.replaceWith(...target.childNodes) removes that marker synchronously and leaves the real children as direct siblings at the render position. The regression test also asserts the button is a direct child of the async block.

File: js/core/src/MountCustomComponent.test.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+	mounted = mount(MountCustomComponent, {
+		target: document.body,
+		props: {
+			node: {
+				component: Promise.resolve({ default: AsyncChildren }),
+				runtime: Promise.resolve({
+					createRawSnippet: runtime_create_raw_snippet,
+					mount,
+					unmount
```
[the remark this answers is not in the document]
Comment (abidlabs) [reply js/core/src/MountCustomComponent.test.ts]: Addressed in ed5009942. The fixture now loads its component and Svelte exports in an isolated iframe module graph, so it has independent runtime singleton state from the host test. I also validated the inverse: bypassing the bridge makes this test fail with the original Cannot read properties of null (reading nodes) stack; restoring the bridge passes.

**Comment by abidlabs** [comment]
## Before/after Spaces consolidated

The #13131 and #13818 demos were identical, so they now share one before/after pair:

- **Before:** https://huggingface.co/spaces/abidlabs/gradio-13131-before
- **After:** https://huggingface.co/spaces/abidlabs/gradio-13131-after

The pair covers the #13818 remount regression along with async children, nested context propagation, direct-child layout, and the synchronous control case. Full expected results are listed in the PR body and the [shared verification comment](https://github.com/gradio-app/gradio/pull/13817#issuecomment-5517281037).

**Comment by iwr-redmond** [comment]
@abidlabs you may wish to test the PR against the detailed reproduction provided by @Col0ring [here](https://github.com/Col0ring/gradio6.9-dynamic-children-rendering-issue).

File: js/core/src/MountCustomComponent.svelte [hunk 3]
```diff
@@ -19,15 +23,36 @@
...
+		// Recreate the snippet in the custom component's Svelte runtime, then
+		// replace its temporary marker with children from Gradio's runtime.
+		const runtime_children =
+			children && _runtime.createRawSnippet
+				? _runtime.createRawSnippet(() => ({
+						render: () => "<span hidden></span>",
+						setup: (target) => {
+							const mounted_children = mount(MountChildren, {
+								target,
+								props: { children }
```
Comment (Col0ring) [inline js/core/src/MountCustomComponent.svelte]: Hi @abidlabs! Thanks for this fix — it does unblock rendering. But I think there is still a piece missing around here: nested custom components lose their Svelte context.

Since every custom component is mounted as its own root, `getContext` inside a nested custom component never reaches the ancestor custom component:

- the ancestor's context lives in the **custom component's** runtime, not Gradio's, so it is not visible from the host side;
- `setContext`/`getContext` are copy-on-write (`component_context.c ??= new Map(get_parent_context(...))`), so nothing a component writes into its context can be read back from the host afterwards.

So as things stand, nothing is handed to `mount()`'s `context` option and the two runtimes' context chains stay completely disconnected. I think support for propagating context needs to be added here as well: the ancestor custom component's context has to be handed down to the nested mounts, so that `setContext`/`getContext` keep working across custom component boundaries.

Comment (abidlabs) [reply js/core/src/MountCustomComponent.svelte]: Thanks @iwr-redmond @Col0ring for the feedback! I'm still reviewing eveyrthing but this should be fixed at least

File: js/core/src/MountCustomComponent.svelte
Comment (Col0ring) [inline js/core/src/MountCustomComponent.svelte]: @abidlabs There's another issue here: 

`<span bind:this={el}></span>` stays in the DOM as a permanent wrapper around every mounted custom component, with its default `display: inline`. Whenever a custom component sits inside a container that lays out its children (a flex row, a grid, a custom layout component), the span becomes the item instead of the component: the parent's `gap` / `align-items` / `flex` apply to the span, the span shrink-to-fits so everything from it downwards measures 0 wide, and percentage sizes inside the component resolve against the wrong box.

Interesting contrast within this PR: the temporary children target _is_ unwrapped with `target.replaceWith(...target.childNodes)`, while the component's own mount target is kept.

Neutralising those spans from the console restores the layout exactly, with no other change:

```js
[...document.querySelectorAll('span:not([class]):not([style])')]
  .filter((s) => s.children.length)
  .forEach((s) => (s.style.display = 'contents'));
```

So the minimal fix looks like:

```svelte
<span bind:this={el} style="display: contents"></span>
```

It has to be `display: contents` rather than the same `replaceWith` treatment: the effect mounts into `el` again on every re-run, so the span needs to stay in the document.

Comment (abidlabs) [reply js/core/src/MountCustomComponent.svelte]: Thanks @Col0ring for catching this! Fixed, as you can now confirm with the 2nd tab here: https://huggingface.co/spaces/abidlabs/gradio-13131-after

I'll do some more testing and then open up this PR for review officially later today. If you can do another pass after that, would appreciate!

Comment (Col0ring) [reply js/core/src/MountCustomComponent.svelte]: Thanks for the quick fix! One more thought about this: semantically `<span>` is a text-level element, while this one is purely an implementation detail — and it means any `span { ... }` rule from a theme or `gr.Blocks(css=...)` lands on every custom component wrapper in the app. A dedicated tag might be a better fit here:

```svelte
<gradio-custom-component bind:this={el} style="display: contents"
></gradio-custom-component>
```

**Comment by abidlabs** [comment]
Thanks for the consistent feedback and suggestions @Col0ring! PR should be ready now, opening it up for the rest of the team to take a look at.

All the fixes can be seen in this Space: https://huggingface.co/spaces/abidlabs/gradio-13131-after

**Comment by iwr-redmond** [comment]
Is the reactive update mount count supposed to increase by increments of two?

[](https://github.com/user-attachments/assets/9c3339e3-2635-498e-86dc-90a372e9c237)

**Comment by abidlabs** [comment]
Thanks @iwr-redmond , this is fixed now and I've updated the Space. Please let me know if you see any other unexpected behavior

**Comment by iwr-redmond** [comment]
Thanks @abidlabs, this fix has also improved the speed of the demo.

File: js/core/src/MountCustomComponent.svelte [hunk 4]
```diff
@@ -1,38 +1,92 @@
...
 			unmount: typeof import("svelte").unmount;
+			createRawSnippet?: typeof import("svelte").createRawSnippet;
+			getAllContexts?: typeof import("svelte").getAllContexts;
 		}
 	);
 	let el: HTMLElement | null = $state(null);
 
 	$effect(() => {
 		if (!el || !runtime || !component) return;
+		const target = el;
```
Comment (dawoodkhan82) [inline js/core/src/MountCustomComponent.svelte:33]:
```suggestion
		const target = el;
		void node.id;
```

`@gr.render` re-renders reuse this instance for a node with a new id, so without tracking `node.id` the custom component never remounts and stays registered under the old id.

**Review by dawoodkhan82 — APPROVED** [review]
besides the small `gr.render` issue. lgtm!

**Comment by abidlabs** [comment]
Thanks so much for reviewing @Col0ring @iwr-redmond @dawoodkhan82!
